--- a/net/minecraft/world/level/pathfinder/PathTypeCache.java
+++ b/net/minecraft/world/level/pathfinder/PathTypeCache.java
@@ -6,10 +_,10 @@
 import org.jspecify.annotations.Nullable;
 
 public class PathTypeCache {
-    private static final int SIZE = 4096;
-    private static final int MASK = 4095;
-    private final long[] positions = new long[4096];
-    private final PathType[] pathTypes = new PathType[4096];
+    private static final int SIZE = 16384;
+    private static final int MASK = SIZE - 1;
+    private final long[] positions = new long[SIZE];
+    private final PathType[] pathTypes = new PathType[SIZE];
 
     public PathType getOrCompute(BlockGetter level, BlockPos pos) {
         long packedBlockPos = pos.asLong();
